src/device/mod.rs

```diff
@@ -0,0 +1,289 @@
+    /// for an invalid selection.
+    fn queue_mut(&mut self, index: u16) -> Option<&mut Queue<M>>;
+
+    /// Return the features advertised by the device.
+    ///
+    /// Using `u64` for the entire feature set because it's wide enough for the entire feature
+    /// set specified by the standard (and we can later switch to `u128` if necessary).
+    fn device_features(&self) -> u64;
+
+    /// Return the features bits written by the driver.
+    fn driver_features(&self) -> u64;
+
+    /// Set the specified page of features written by the driver.
+    fn set_driver_features(&mut self, page: u32, value: u32);
+
+    /// Return the current device status flags.
+    fn device_status(&self) -> u8;
+
+    /// Validate the current device status with respect to a group of flags that must be set,
+    /// and another group that must be cleared.
+    fn check_device_status(&self, set: u8, cleared: u8) -> bool {
+        self.device_status() & (set | cleared) == set
+    }
+
+    /// Set the value of the device status flags. This is just a simple accessor method, and
+    /// `ack_device_status` must be used to process a status update coming from the driver.
+    fn set_device_status(&mut self, status: u8);
+
+    /// Acknowledge a status update from the driver, based on the provided value. This method
+    /// is not just a simple accessor, but rather is expected to handle virtio device status
+    /// transitions (which may involve things such as calling activation or reset logic).
+    fn ack_device_status(&mut self, status: u8) {
+        use status::*;
+        let current_status = self.device_status();
+
+        // Match changed bits.
+        match !current_status & status {
+            ACKNOWLEDGE if current_status == RESET => {
+                self.set_device_status(status);
+            }
+            DRIVER if current_status == ACKNOWLEDGE => {
+                self.set_device_status(status);
+            }
+            FEATURES_OK if current_status == (ACKNOWLEDGE | DRIVER) => {
+                // Ensure the driver did not ack a feature that wasn't actually offered. If that
+                // happens, the standard specifies devices should not actually accept/set the
+                // `FEATURES_OK` status bit.
+                if !self.device_features() & self.driver_features() != 0 {
+                    warn!("driver accepted invalid feature bits");
+                    return;
+                }
+
+                // TODO: Bring this (and other feature definitions) to the vm-virtio crate proper.
+                // Using a local const temporarily until then.
+                const VIRTIO_F_RING_EVENT_IDX: u64 = 29;
+
+                // Set the appropriate configuration flag for all queues if we offered the
+                //`VIRTIO_F_RING_EVENT_IDX` feature and the driver acknowledged it.
+                if self.driver_features() & (1 << VIRTIO_F_RING_EVENT_IDX) != 0 {
+                    for i in 0..self.num_queues() {
```

Comment:
I think there's a way to test this too with `Dummy` though I don't have the specifics.